### DISCUSSION TITLE
Développement (FR) // Development (EN)

### DISCUSSION THREAD

**Discussion by Dadoum** [Dadoum/Provision#4] [Announcements]
Here you can talk @tkreind

**Comment by Dadoum** [comment]
@BEEDELLROKEJULIANLOCKHART

**Reply by roke-julian-lockhart** [reply]
J'essaierai de contribuer davantage quand j'aurai appris la syntaxe du C.

**Comment by Dadoum** [comment]
Discord: Dadoum#9690

**Reply by lonkelle** [reply]
Added you on Discord!

**Comment by Dadoum** [comment]
@iMonZ

**Reply by iMonZ** [reply]
Thanks! I’m looking into it later

**Reply by iMonZ** [reply]
@Dadoum hey what are you actually doing with the android stuff here?

**Reply by Dadoum** [reply]
@iMonZ To install apps, you need a developer certificate, to get one you need to authenticate to apple servers, and to authenticate you need a set of headers called "anisette". And the way to get them are non trivial, and difficult to reverse engineer (this is related to fair-play encryption, so if we crack it, we potentially open the door to cracking the whole iTunes library). Instead, we use Apple librairies. On Windows, we can get them from iTunes and iCloud but on Linux, this requires trickery. Apple never released Linux software, but it already released Android software. Android libraries are different from Linux ones but libhybris allows us to load them pretty easily (if you know how to use it). The second difficulty is using libraries; no documentation. We are here.

**Reply by iMonZ** [reply]
Thanks for letting me know. Now everything make sense again!

**Comment by Dadoum** [comment]
Ça avance !

**Comment by lonkelle** [comment]
@Dadoum Looks like you've gone back and forth between CoreADI.dll and CoreADI for Android but settled on the Android lib since this repo is focused on the CoreADI Android implementation in Apple Music. 

So, tell me, how far have you got and what's your next step? PS. What version of Linux do you run?

**Reply by Dadoum** [reply]
Pretty much all (high-level) APIs of Apple Music requires an object named RequestContext. Its initialization is tricky but two days ago I finished the initialization phase. Now I am exploring all APIs in "libstoreservicescore" and "libandroidappmusic" to see where I should continue.

**Reply by Dadoum** [reply]
(I run Arch Linux with LTS kernel, x86_64 but should support x86, armv7 and aarch64)

**Reply by lonkelle** [reply]
My money is on `libstoreservicescore`. Hell yeah! Nice work initializing the object already. TWO DAYS AGO is recent which means you can play with the API directly; you could figure this out soon. And honestly, thanks to APKMirror - the version of the APK we need will always be available.

Since you're actually near the end of finding out if this is possible, I can help be a sounding board. In the original issue we spoke on, you said you'd already been able to generate anisette headers via Wine (https://cloudup.com/cE9sZqASAlR) so I'm curious what stopped you going in the "Wine" direction?

**Reply by Dadoum** [reply]
My "minimal" wine AppImage was taking 500 MB and was requiring both 32 bits x86 support and GNUTLS installed in 32 bits. Also, since we can't redistribute iTunes and iCloud library, the AppImage must be crafted with helper script and that requires a lot of useless downloads + more requirements.

When I succeeded to use libhybris, it was originally to RE Anisette, since you can replace gradually function using hook API. But now I abandoned the idea since it's really hard while we can use libraries yet. Also, if we don't reimplement FairPlay we are safer (FairPlay is used everywhere (especially DRMs) by Apple).

**Reply by Dadoum** [reply]
(I accepted your friend request on discord btw)

**Reply by iMonZ** [reply]
> My "minimal" wine AppImage was taking 500 MB and was requiring both 32 bits x86 support and GNUTLS installed in 32 bits. Also, since we can't redistribute iTunes and iCloud library, the AppImage must be crafted with helper script and that requires a lot of useless downloads + more requirements.
> 
> When I succeeded to use libhybris, it was originally to RE Anisette, since you can replace gradually function using hook API. But now I abandoned the idea since it's really hard while we can use libraries yet. Also, if we don't reimplement FairPlay we are safer (FairPlay is used everywhere (especially DRMs) by Apple).

Maybe Snapcraft can help with that?
Should be easier than AppImages and much more convenient.

**Comment by Dadoum** [comment]
I write here some of my findings:

The class "AnisetteProtocolAction" contains an interesting method "_provisionWithContext"; it is a non-static function taking RequestContext - we **have** a RequestContext

But to try this route, we need to create instance of AnisetteProtocolAction, a constructor that takes a string and a CFData. Since we don't know what is exactly the argument, we need either to make guesses (which seems to me too complex), or we see elsewhere for its creation. Here is what if found (at the bottom):

////////////////////////////////////////////////

J'écris ici mes recherches:

La classe "AnisetteProtocolAction" contient la méthode "_provisionWithContext" qui vu son nom pourrait nous aider: elle n'est pas statique et prend un objet RequestContext, objet que l'on sait déjà créer dans Provision.

Cependant, il faut quand même créer une instance du type AnisetteProtocolAction, et cela requiert un std::string et un CFData, sans plus d'indications. Soit on essaie de déduire leur contenu (ce qui me semble compliqué), soit on voit autre chose qui pourrait potentiellement appeler cette fonction. Voici ce que j'ai trouvé:

![image](https://user-images.githubusercontent.com/24679280/125487123-6d69d163-c9fa-4469-bb2b-31540d20d4b8.png)

**Reply by lonkelle** [reply]
Okay, so my findings so far: `cvu8io98wun` is used to obtain the version / protocol data, and `vdfut768ig` is the main worker function.

Trying to line this up with CoreADI.dll, we have: 
• `ADIProvisioningStart` (maybe libCoreADI.so version of `provisionWithContext`) - which takes a Server PIM (`SPIM`) and returns a Client PIM (`CPIM`) and `session ID`. 
• `ADISynchronize` (probs libCoreADI.so version of `syncMachineWithContext`)   - which takes Server Intermediate Metadata (`SIM`) and returns Synchronization Resume Metadata (`SRM`).

Missing? `ADIProvisioningEnd` which takes a server's Persistent Token Metadata (`PTM`) and Trust Key (`TK`) and writes provisioning data to disk. `actionForHeaders` is a really exciting name. We just need find a way to initialize `AnisetteProtocolAction`.

**Reply by lonkelle** [reply]
Maybe `cvu8io98wun` can be used to get the initialization data we need given it's what's used to return the protocol data?

**Reply by Dadoum** [reply]
I'm not even sure if actionForHeaders requires an AnisetteProtocolAction instance. Since it calls the constructor inside the function. Just need to create a multimap from D to try this out

**Reply by lonkelle** [reply]
My experience is with `CoreADI.dll` - so I'm doing my best to line everything up. I'd be surprised if we could get them directly like this but SUPER EXCITED to see **if** it returns what we need. Anisette is so close!
